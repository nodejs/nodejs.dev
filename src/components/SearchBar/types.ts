import type { SerialisedIndexData } from 'elasticlunr';
import type { SearchResult } from '../../types';

export interface SearchProps {
  localSearchLearnPages: {
    index: SerialisedIndexData<SearchResult>;
  };
}
