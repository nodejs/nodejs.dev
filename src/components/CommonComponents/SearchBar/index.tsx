import React, { useMemo, useEffect, useState, createRef } from 'react';
import { FormattedMessage } from 'react-intl';
import { LocalizedLink as Link } from 'gatsby-theme-i18n';
import TravelExploreIcon from '@mui/icons-material/TravelExplore';
import CloseIcon from '@mui/icons-material/Close';
import classNames from 'classnames';
import { AnimatePresence, motion } from 'framer-motion';
import { useClickOutside } from 'react-click-outside-hook';
import SectionTitle from '../SectionTitle';
import useKeyPress from '../../../hooks/useKeyPress';
import { useSearchResults } from '../../../hooks/useSearchResults';
import { SearchResult } from '../../../types';
import styles from './index.module.scss';

const containerTransition = { type: 'spring', damping: 22, stiffness: 150 };
const containerVariants = {
  expanded: {
    minHeight: '5em',
    width: '100%',
    maxWidth: '450px',
    boxShadow: '0px 2px 12px 3px rgba(153, 204, 125, 0.14)',
  },
  collapsed: {
    minHeight: '0em',
    width: '100%',
    maxWidth: '100px',
    boxShadow: 'none',
  },
};

const MotionCloseIcon = motion(CloseIcon);

const SearchBar = (): JSX.Element => {
  const [query, setQuery] = useState('');
  const searchInputRef = createRef<HTMLInputElement>();

  const [isExpanded, setExpanded] = useState(false);
  const [parentRef, isClickedOutside] = useClickOutside();

  const search = useSearchResults();

  const results = useMemo(() => search(query), [query, search]);

  const isEmpty = !results || results.length === 0;

  const changeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
    e.preventDefault();

    setQuery(e.target.value);
  };

  const expandContainer = () => {
    setExpanded(true);
  };

  const collapseContainer = () => {
    setExpanded(false);
    setQuery('');
  };

  useEffect(() => {
    if (isClickedOutside) {
      collapseContainer();
    }
  }, [isClickedOutside]);

  const containerClassNames = classNames(styles.searchBarContainer, {
    [styles.expanded]: isExpanded,
  });

  const onKeyPressHandler = () => {
    if (!isExpanded) {
      expandContainer();
    }

    if (searchInputRef.current) {
      searchInputRef.current.focus();
    }
  };

  useKeyPress('/', () => onKeyPressHandler());
  useKeyPress('Escape', () => {
    if (isExpanded) {
      collapseContainer();
      searchInputRef.current?.blur();
    }
  });

  return (
    <motion.div
      className={containerClassNames}
      animate={isExpanded ? 'expanded' : 'collapsed'}
      initial="collapsed"
      variants={containerVariants}
      transition={containerTransition}
      ref={parentRef}
    >
      <div
        className={styles.searchInputContainer}
        onKeyPress={onKeyPressHandler}
        onClick={onKeyPressHandler}
        role="presentation"
      >
        <TravelExploreIcon className={styles.searchIcon} />
        <label htmlFor="searchInput">
          <span>
            {!isExpanded && (
              <FormattedMessage id="components.searchBar.placeholder" />
            )}
          </span>
          <input
            ref={searchInputRef}
            autoComplete="off"
            className={styles.inputText}
            id="searchInput"
            name="query"
            type="text"
            value={query}
            onFocus={onKeyPressHandler}
            onChange={changeHandler}
          />
        </label>
        <AnimatePresence>
          {isExpanded && (
            <MotionCloseIcon
              className={styles.closeIcon}
              key="close-icon"
              initial={{ opacity: 0 }}
              animate={{ opacity: 1 }}
              exit={{ opacity: 0 }}
              onClick={collapseContainer}
              transition={{ duration: 0.2 }}
            />
          )}
        </AnimatePresence>
      </div>

      {isExpanded && (
        <div className={styles.searchContent}>
          {isEmpty && (
            <div className={styles.loadingWrapper}>
              <div className={styles.warningMessage}>
                <FormattedMessage
                  id={
                    query.length
                      ? 'components.searchBar.search.noResults'
                      : 'components.searchBar.search.title'
                  }
                />
              </div>
            </div>
          )}
          {!isEmpty && (
            <ul>
              {results.map((result: SearchResult) => {
                const sectionPath =
                  result.category === 'api'
                    ? ['home', result.category, result.title]
                    : ['home', result.category];

                const displayTitle = result.displayTitle || result.title;

                return (
                  <li key={result.id}>
                    <Link to={result.slug}>
                      {(result.wrapInCode && <code>{displayTitle}</code>) || (
                        <span>{displayTitle}</span>
                      )}
                    </Link>
                    <SectionTitle path={sectionPath} />
                  </li>
                );
              })}
            </ul>
          )}
        </div>
      )}
    </motion.div>
  );
};

export default SearchBar;
