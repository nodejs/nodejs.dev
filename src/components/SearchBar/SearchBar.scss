.searchBarContainer {
  background-color: var(--color-fill-top-nav);
  border-radius: 6px;
  box-shadow: 0px 2px 12px 3px rgba(153, 204, 125, 0.14);
  display: flex;
  flex-direction: column;
  height: 3.8em;
  width: 34em;
}

.searchInputContainer {
  align-items: center;
  display: flex;
  min-height: 4em;
  padding: 2px 15px;
  position: relative;
  width: 100%;
}

.inputText {
  background-color: transparent;
  border: none;
  border-radius: 6px;
  color: var(--color-text-primary);
  font-size: var(--font-size-body3);
  font-weight: 500;
  height: 100%;
  outline: none;
  width: 100%;
  &:focus {
    outline: none;
    &::placeholder {
      opacity: 0;
    }
  }

  &::placeholder {
    color: var(--color-text-primary);
    transition: all 250ms ease-in-out;
  }
}

.closeIcon {
  position: relative;
  right: 10px;
  top: 5px;
}

.searchIcon {
  color: var(--color-brand-primar);
  font-size: 27px;
  margin-right: 10px;
  margin-top: 6px;
  vertical-align: middle;
}

.searchContent {
  display: flex;
  flex-direction: column;
  height: 100%;
  overflow-y: auto;
  width: 100%;
}

.loadingWrapper {
  align-items: center;
  display: flex;
  height: 100%;
  justify-content: center;
  width: 100%;
}

.warningMessage {
  align-self: center;
  color: var(--color-text-primary);
  display: flex;
  font-size: 14px;
  justify-self: center;
}

.resultItem {
  color: var(--color-text-primary);
  list-style-type: none;
}
