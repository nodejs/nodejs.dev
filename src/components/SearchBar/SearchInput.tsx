import React, { useMemo, useEffect, useState } from 'react';
import { AnimatePresence, motion } from 'framer-motion';
import { useClickOutside } from 'react-click-outside-hook';
import { Index } from 'elasticlunr';
import { Link } from 'gatsby';
import { SearchResult } from '../../types';
import { SearchProps } from './types';

const containerTransition = { type: 'spring', damping: 22, stiffness: 150 };
const containerVariants = {
  expanded: {
    minHeight: '5em',
    width: '100%',
    maxWidth: '30em',
    boxShadow: '0px 2px 12px 3px rgba(153, 204, 125, 0.14)',
  },
  collapsed: {
    minHeight: '3em',
    width: '100%',
    maxWidth: '7em',
    boxShadow: 'none',
  },
};

const SearchInput = ({ localSearchLearnPages }: SearchProps): JSX.Element => {
  const [query, setQuery] = useState('');
  const [results, setResults] = useState<SearchResult[]>([]);
  const [noResults, setNoResults] = useState(false);

  const isEmpty = !results || results.length === 0;
  const [isExpanded, setExpanded] = useState(false);
  const [parentRef, isClickedOutside] = useClickOutside();

  const storeIndex = useMemo(
    () => Index.load<SearchResult>(localSearchLearnPages.index),
    [localSearchLearnPages.index]
  );

  const searchForResults = (currentQuery: string) => {
    const currentResults = storeIndex
      .search(currentQuery, { expand: true })
      .map(({ ref }) => storeIndex.documentStore.getDoc(ref) as SearchResult);

    setResults(currentResults.slice(0, 20));
  };

  const changeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
    e.preventDefault();

    if (e.target.value === '') {
      setNoResults(false);
      setResults([]);
    }

    setQuery(e.target.value);
    searchForResults(e.target.value);
  };

  const expandContainer = () => {
    setExpanded(true);
  };

  const collapseContainer = () => {
    setExpanded(false);
    setQuery('');
    setNoResults(false);
    setResults([]);
  };

  useEffect(() => {
    if (isClickedOutside) {
      collapseContainer();
    }
  }, [isClickedOutside]);

  return (
    <motion.div
      className="searchBarContainer"
      animate={isExpanded ? 'expanded' : 'collapsed'}
      initial="collapsed"
      variants={containerVariants}
      transition={containerTransition}
      ref={parentRef}
    >
      <div className="searchInputContainer">
        <i className="material-icons searchIcon">search</i>
        <input
          autoComplete="off"
          className="inputText"
          name="query"
          value={query}
          onChange={changeHandler}
          placeholder="Search"
          onFocus={expandContainer}
        />
        <AnimatePresence>
          {isExpanded && (
            <motion.span
              className="material-icons closeIcon"
              key="close-icon"
              initial={{ opacity: 0 }}
              animate={{ opacity: 1 }}
              exit={{ opacity: 0 }}
              onClick={collapseContainer}
              transition={{ duration: 0.2 }}
            >
              close
            </motion.span>
          )}
        </AnimatePresence>
      </div>

      {isExpanded && (
        <div className="searchContent">
          {isEmpty && !noResults && (
            <div className="loadingWrapper">
              <div className="warningMessage">Start typing to Search</div>
            </div>
          )}
          {!isEmpty && (
            <ul>
              {results.map((result: SearchResult) => (
                <li className="resultItem" key={result.id}>
                  <Link to={`/learn/${result.slug}`}>{result.title}</Link>
                </li>
              ))}
            </ul>
          )}
        </div>
      )}
    </motion.div>
  );
};

export default SearchInput;
