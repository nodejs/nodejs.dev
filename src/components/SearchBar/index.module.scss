.search-bar-container {
  background-color: var(--color-fill-top-nav);
  border-radius: 6px;
  display: flex;
  flex-direction: column;
  height: fit-content;
  max-height: 90vh;
  overflow-y: auto;

  &.expanded {
    padding-bottom: 10px;
  }

  .search-input-container {
    align-items: center;
    display: flex;
    position: relative;
    width: 100%;

    label {
      align-items: center;
      color: var(--color-text-accent);
      cursor: pointer;
      display: flex;
      font-weight: var(--font-weight-semibold);
      transition: all 250ms ease-in-out;
      width: 100%;
    }

    .input-text {
      background-color: transparent;
      border: none;
      border-radius: 6px;
      color: var(--color-text-primary);
      font-size: var(--font-size-body1);
      font-weight: var(--font-weight-semibold);
      height: 100%;
      outline: none;
      width: 100%;

      &:focus {
        outline: none;
      }
    }

    .close-icon {
      padding: 0 15px 0 0;
    }

    .search-icon,
    .close-icon {
      color: var(--color-text-accent);
      vertical-align: middle;
    }
  }

  .search-content {
    display: flex;
    flex-direction: column;
    height: 100%;
    width: 100%;

    .loading-wrapper {
      align-items: center;
      display: flex;
      height: 100%;
      justify-content: center;
      width: 100%;
    }

    .warning-message {
      align-self: center;
      color: var(--color-text-primary);
      display: flex;
      font-size: 14px;
      justify-self: center;
    }

    ul {
      list-style: circle;
      padding: 0 var(--space-24) 0 var(--space-32);

      li {
        color: var(--color-text-primary);
        display: list-item;
        margin: 0;
        padding: 5px 0;

        > a {
          text-decoration: none;
          text-overflow: ellipsis;
          text-transform: capitalize;
        }
      }
    }
  }
}
