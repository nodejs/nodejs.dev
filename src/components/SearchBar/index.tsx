import React, { useState } from 'react';
import { useFlexSearch } from 'react-use-flexsearch';
import { Formik, Form, Field } from 'formik';
import { graphql, Link, useStaticQuery } from 'gatsby';

const SearchBar = (): JSX.Element => {
  const queryData = useStaticQuery(graphql`
    query {
      localSearchLearnPages {
        id
        store
        index
      }
    }
  `);
  const index = queryData.localSearchLearnPages.index;
  const store = queryData.localSearchLearnPages.store;
  const [query, setQuery] = useState('');

  const results = useFlexSearch(query, index, store);
  console.log('results', results);
  return (
    <div>
      <Formik
        initialValues={{ query: '' }}
        onSubmit={(values, { setSubmitting }) => {
          setQuery(values.query);
          setSubmitting(false);
        }}
      >
        <Form>
          <Field name="query" />
        </Form>
      </Formik>
      <ul>
        {results.map(
          (result: {
            id: React.Key | null | undefined;
            slug: string;
            title:
              | boolean
              | React.ReactChild
              | React.ReactFragment
              | React.ReactPortal
              | null
              | undefined;
          }) => (
            <li key={result.id}>
              <Link to={`learn/${result.slug}`}>{result.title}</Link>
            </li>
          )
        )}
      </ul>
    </div>
  );
};
export default SearchBar;
