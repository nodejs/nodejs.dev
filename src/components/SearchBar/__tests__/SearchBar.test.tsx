import React from 'react';
import { render } from '@testing-library/react';
import SearchBar from '..';

describe('Searbar component', () => {
  it.only('should render correctly', () => {
    const component = render(<SearchBar />);
    expect(component).toMatchSnapshot();
  });
});
