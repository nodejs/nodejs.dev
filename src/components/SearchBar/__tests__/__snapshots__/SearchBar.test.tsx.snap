// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearInput component should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="searchBarContainer"
        style="min-height: 0em; width: 100%; max-width: 100px; box-shadow: none;"
      >
        <div
          class="searchInputContainer"
          role="presentation"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium searchIcon css-i4bv87-MuiSvgIcon-root"
            data-testid="TravelExploreIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19.3 16.9c.4-.7.7-1.5.7-2.4 0-2.5-2-4.5-4.5-4.5S11 12 11 14.5s2 4.5 4.5 4.5c.9 0 1.7-.3 2.4-.7l3.2 3.2 1.4-1.4-3.2-3.2zm-3.8.1c-1.4 0-2.5-1.1-2.5-2.5s1.1-2.5 2.5-2.5 2.5 1.1 2.5 2.5-1.1 2.5-2.5 2.5zM12 20v2C6.48 22 2 17.52 2 12S6.48 2 12 2c4.84 0 8.87 3.44 9.8 8h-2.07c-.64-2.46-2.4-4.47-4.73-5.41V5c0 1.1-.9 2-2 2h-2v2c0 .55-.45 1-1 1H8v2h2v3H9l-4.79-4.79C4.08 10.79 4 11.38 4 12c0 4.41 3.59 8 8 8z"
            />
          </svg>
          <label
            for="searchInput"
          >
            <span>
              Search
            </span>
            <input
              autocomplete="off"
              class="inputText"
              id="searchInput"
              name="query"
              type="text"
              value=""
            />
          </label>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="searchBarContainer"
      style="min-height: 0em; width: 100%; max-width: 100px; box-shadow: none;"
    >
      <div
        class="searchInputContainer"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium searchIcon css-i4bv87-MuiSvgIcon-root"
          data-testid="TravelExploreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19.3 16.9c.4-.7.7-1.5.7-2.4 0-2.5-2-4.5-4.5-4.5S11 12 11 14.5s2 4.5 4.5 4.5c.9 0 1.7-.3 2.4-.7l3.2 3.2 1.4-1.4-3.2-3.2zm-3.8.1c-1.4 0-2.5-1.1-2.5-2.5s1.1-2.5 2.5-2.5 2.5 1.1 2.5 2.5-1.1 2.5-2.5 2.5zM12 20v2C6.48 22 2 17.52 2 12S6.48 2 12 2c4.84 0 8.87 3.44 9.8 8h-2.07c-.64-2.46-2.4-4.47-4.73-5.41V5c0 1.1-.9 2-2 2h-2v2c0 .55-.45 1-1 1H8v2h2v3H9l-4.79-4.79C4.08 10.79 4 11.38 4 12c0 4.41 3.59 8 8 8z"
          />
        </svg>
        <label
          for="searchInput"
        >
          <span>
            Search
          </span>
          <input
            autocomplete="off"
            class="inputText"
            id="searchInput"
            name="query"
            type="text"
            value=""
          />
        </label>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
